core/src/apps/cardano/auxiliary_data.py

```diff
@@ -0,0 +1,252 @@
+from trezor.crypto import hashlib
+from trezor.crypto.curve import ed25519
+from trezor.messages import (
+    CardanoAddressType,
+    CardanoAuxiliaryDataType,
+    CardanoMetadataType,
+)
+
+from apps.common import cbor
+
+from ..common.seed import remove_ed25519_prefix
+from .address import derive_address_bytes, derive_human_readable_address
+from .helpers import INVALID_AUXILIARY_DATA, bech32
+from .helpers.bech32 import HRP_JORMUN_PUBLIC_KEY
+from .helpers.paths import SCHEMA_STAKING_ANY_ACCOUNT
+from .layout import confirm_catalyst_registration, show_auxiliary_data_hash
+
+if False:
+    from typing import Dict, List, Optional, Tuple, Union
+    from trezor import wire
+
+    from trezor.messages.CardanoCatalystRegistrationParametersType import (
+        CardanoCatalystRegistrationParametersType,
+    )
+    from trezor.messages.CardanoTxAuxiliaryDataType import CardanoTxAuxiliaryDataType
+
+    CatalystRegistrationPayload = Dict[int, Union[bytes, int]]
+    CatalystRegistrationSignature = Dict[int, bytes]
+    CatalystRegistration = Dict[
+        int, Union[CatalystRegistrationPayload, CatalystRegistrationSignature]
+    ]
+
+    from . import seed
+
+AUXILIARY_DATA_HASH_SIZE = 32
+PUBLIC_KEY_LENGTH = 32
+VOTING_REGISTRATION_HASH_SIZE = 32
+CATALYST_REGISTRATION_KEY = 61284
+CATALYST_REGISTRATION_SIGNATURE_KEY = 61285
+
+
+def validate_auxiliary_data(
+    auxiliary_data: Optional[CardanoTxAuxiliaryDataType],
+) -> None:
+    if not auxiliary_data:
+        return
+
+    if auxiliary_data.type == CardanoAuxiliaryDataType.BLOB:
+        if not auxiliary_data.blob:
+            raise INVALID_AUXILIARY_DATA
+
+        _validate_auxiliary_data_blob(auxiliary_data.blob)
+    elif auxiliary_data.type == CardanoAuxiliaryDataType.TUPLE:
+        # currently we only support CATALYST_REGISTRATION
+        if (
+            auxiliary_data.blob
+            or not auxiliary_data.metadata
+            or auxiliary_data.metadata.type != CardanoMetadataType.CATALYST_REGISTRATION
+        ):
+            raise INVALID_AUXILIARY_DATA
+
+        _validate_catalyst_registration_parameters(
+            auxiliary_data.metadata.catalyst_registration_parameters
+        )
+    else:
+        raise INVALID_AUXILIARY_DATA
+
+
+def _validate_auxiliary_data_blob(auxiliary_data: bytes) -> None:
```

Comment:
`auxiliary_data_blob` perhaps to differentiate it from `auxiliary_data` top-level structure

Reply:
https://github.com/vacuumlabs/trezor-firmware/pull/57/commits/f73f563cd7b142506f75b45773fe521898ae04ab